src/CircularArrayBuffers.jl

```diff
@@ -76,24 +76,42 @@ capacity(cb::CircularArrayBuffer{T,N}) where {T,N} = size(cb.buffer, N)
 isfull(cb::CircularArrayBuffer) = cb.nframes == capacity(cb)
 Base.isempty(cb::CircularArrayBuffer) = cb.nframes == 0
 
+"""
+    _buffer_index(cb::CircularArrayBuffer, i::Int)
+
+    Return the index of the `i`-th element in the buffer.
+"""
 @inline function _buffer_index(cb::CircularArrayBuffer, i::Int)
-    ind = (cb.first - 1) * cb.step_size + i
-    if ind > length(cb.buffer)
-        ind - length(cb.buffer)
+    idx = (cb.first - 1) * cb.step_size + i
+    return wrap_index(idx, length(cb.buffer))
+end
+@inline _buffer_index(cb::CircularArrayBuffer, I::AbstractVector{<:Integer}) = map(Base.Fix1(_buffer_index, cb), I)
+
+"""
+    wrap_index(idx, n)
+
+    Return the index of the `idx`-th element in the buffer, if index is one past the size, return 1, else error.
+"""
+function wrap_index(idx, n)
+    if idx <= n
+        return idx
+    elseif idx <= 2n
+        return idx - n
     else
-        ind
+        @info "oops! idx $(idx) > 2n $(2n)"
+        return idx - n
     end
 end
-@inline _buffer_index(cb::CircularArrayBuffer, I::AbstractVector{<:Integer}) = map(Base.Fix1(_buffer_index, cb), I)
 
+"""
+    _buffer_frame(cb::CircularArrayBuffer, i::Int)
+
+    Return the index of the `i`-th frame in the buffer.
```

Comment:
```suggestion
Here `i` is assumed to be the last dimension of `cb`. Each `frame` means a slice of the last dimension. Since we use *circular frames* (the `data` buffer) underlying, this function transforms the logical `i`-th frame to the real frame of the internal buffer.
```